[MRG+1]: Add decision tree plotting.

nvdv opened the pull request:
Here's initial version of decision tree plot function (#6261). 
Tell me what you think and I'll add all docs then. Also I am thinking about adding `matplotlib` plotting to this function.

Titan-C committed:
[MAINT] Install sphinx-gallery v0.1.0

Titan-C committed:
conf.py integrate sphinx-gallery

Titan-C committed:
update image sources namespace

Titan-C committed:
update references link namespace

Titan-C committed:
update reference namespace in gallery scripts

Titan-C committed:
Add carousel

Fix reference link

Titan-C committed:
References to examples

Titan-C committed:
more examples references

Titan-C committed:
[MAINT] Update to sphinx-gallery v0.1.1

Titan-C committed:
Fix example to use rst syntax in notebook comments

Titan-C committed:
Extra missing links complained by sphinx

nvdv committed:
Add decision tree plotting.

nvdv committed:
Add decision tree plotting.

rvraghav93 commented on sklearn/tree/export.py:
```diff
@@ -412,3 +412,33 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+         feature_names=None, class_names=None, label='all',
+         filled=False, leaves_parallel=False, impurity=True,
+         node_ids=False, proportion=False, rotate=False,
+         rounded=False, special_characters=False):
```
Could we just have `*args, **kwargs` and pass these to `export_graphwiz`?

rvraghav93 replied:
Or maybe this is explicit and better... No strong opinion...

Titan-C committed:
[MAINT] update to sphinx-gallery v0.1.2

Titan-C committed:
Remove custom touch to backreferences files

This is now a default in sphinx-gallery v0.1.2

rvraghav93 commented:
Could we have a common `export_tree` / `export` function and have an attribute like `format={'dot'|'ipython_image', default 'dot'}` and rename the `out_file` attribute to `out` modifying its functionality to write to file if a string file name / writable file object is given or `None` to return the dot graph string / `IPython.display.Image` object?

(If we want it named as `export`, it should be done after moving the public `export.py` module to private `_export.py`)

What I am suggesting is something like

```
>>> export(dtc, ..., format='dot', out='None')
graph G {
...
}
```

Or in a jupyter notebook

```
export(dtc, ..., format='jupyter_image', out='None')
<The tree image>
```

Or write to an image file directly
```
export(dtc, ..., format='image', out='tree.png')
```

And of course retain the old functionality (without deprecating as dot files are indeed useful as such)
```
export(dtc, ..., format='dot', out='tree.dot')
```

Titan-C committed:
Sphinx-Gallery has to be loaded after autodoc

Sphinx loads its extensions sequentially as they are specified on the
list. Sphinx-Gallery relies en autodoc to generated the example
thumbnails when auto documenting the API. If autodoc is not loaded
sphinx-gallery skips this step, as it implies the user might not want
it. In this case we need it, so the gallery has to be loaded after
autodoc and autosummary.

The autosummary_generate variable is repeated

Titan-C committed:
Python 2 backwards compatibility in example

This lines breaks the gallery build in python 2, which is the version on
which CircleCI runs. Other option is to build the docs on Python 3

nvdv commented:
anyone?

jmschrei commented:
I was a fan of tree.plot(). What happened to that idea?

rvraghav93 commented:
You mean to say `DecisionTreeClassifier().plot` or `tree.plot` instead of `tree.export(... format='jupyter_image')`?

jmschrei commented:
`DecisionTreeClassifier/Regressor().plot()`

GaelVaroquaux commented:
DecisionTreeClassifier/Regressor().plot() would open a fairly wide modification in the API and what is acceptable in the scope of scikit-learn.

IMHO I needs to go through a full RFC, and enhancement proposal:
https://github.com/scikit-learn/enhancement_proposals

nvdv commented:
I don't have strong opinion on ```DecisionTreeClassifier/Regressor().plot()```  vs ```export.plot(tree)```.
The second one is easier to implement and the first one looks prettier.

nvdv commented:
Any opinions regarding going/not going through proposal?

nvdv commented:
Any opinions regarding going/not going through proposal?

nvdv commented:
Any opinions regarding going/not going through proposal?

rvraghav93 commented:
I think you can go ahead with the `export.plot` one? And later if needed move and bind it with the `tree` class?

rvraghav93 commented:
> DecisionTreeClassifier/Regressor().plot() would open a fairly wide modification in the API and what is acceptable in the scope of scikit-learn.

I missed this comment. I think `export.plot` is the way to go here!

nelson-liu committed:
doc: fix spellings of 'license'

nvdv commented:
Hi, I removed custom import exception and added docstring to plot function in this PR.

nvdv commented:
Any updates on this one?

rvraghav93 commented:
This looks nice to me, but we would want to support plotting, even when run outside ipython (Jupyter) notebook. So I feel an `out='jupyter'/'matplotlib'` param could be useful?

If so make `out='matplotlib'` as the default.

To showcase this, add an example using the new plot function. (That would also function as a nice test)

nvdv commented:
I can add code to detect environment automatically. 
Something like:
```
try:
   from IPython.display import Image
   # Try to plot in Jupyter environment
except ImportError:
   # Try to plot using matplotlib
```

raghavrv commented:
That looks really cool, but I think explicit is better than implicit... Let's have a out param defaulting to matplotlib. Infact if the user has `%matplotlib inline` declared elsewhere in their notebook, it would work as good as `out='jupyter'` ;)

nvdv commented:
I've added `matplotlib` plotting and example (it's pretty basic, but I can change is if it's necessary).
Let me know what you think. 
Thanks!

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        rounded=rounded, special_characters=special_characters)
+    src = graphviz.Source(in_memory_dot_file.getvalue())
+    image = src.pipe(format='png')
+    if out == 'jupyter':
```
Sorry for the nitpicking, but could you validate this out param at the start (before exporting) pl?

nvdv replied:
Done. Thanks!

tguillemot replied:
Ok for me the option jupyter it's not really necessary and complexify the use of that function.
It's better to remove it and let the user put a matplotlib inline into jupyter.

amueller replied:
yeah maybe just that.

raghavrv commented:
Looks good to me, besides the minor nitpick!!

raghavrv commented:
You could rename this to MRG, once you address the nitpick, I'll give it another pass! Thanks!

Also ping @amueller @agramfort for another review...

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+            from matplotlib import pyplot
+        except ImportError:
+            raise ImportError(
+                u'Plotting to Jupyter Notebook requires matplotlib.')
```
Why `u`nicode?

nvdv replied:
Done.

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+            from IPython.display import Image
+        except ImportError:
+            raise ImportError(
+                u'Plotting to Jupyter Notebook requires Jupyter Notebook.')
```
same comment here...

nvdv replied:
Done.

raghavrv replied:
That was fast ;)

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        Options include 'matplotlib' to plot using matplotlib library,
+        'jupyter' to plot to Jupyter Notebook environment.
+    """
+    if out not in {'matplotlib', 'jupyter'}:
```
`{`/`}` --> `(`/`)`

nvdv replied:
yep, forgot about Python 2.6

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        'jupyter' to plot to Jupyter Notebook environment.
+    """
+    if out not in {'matplotlib', 'jupyter'}:
+        raise RuntimeError(
```
ValueError

nvdv replied:
Done

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+    """
+    if out not in {'matplotlib', 'jupyter'}:
+        raise RuntimeError(
+            "Incorrect 'out' value. Use 'jupyter' or 'matplotlib'.")
```
"Incorrect 'out' value passed (%s). Use 'jupyter' or 'ma..." % out

nvdv replied:
Done.

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -412,3 +413,107 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+            from matplotlib import pyplot
+        except ImportError:
+            raise ImportError(
+                'Plotting to Jupyter Notebook requires matplotlib.')
```
Plotting using the option `out='matplotlib'` ;)

nvdv replied:
Good catch. Thanks!

raghavrv commented:
Also could you rebase on master and force push once?

nvdv commented:
Done.

raghavrv commented:
Thanks! LGTM you may now make it `[MRG+1]`.

raghavrv commented:
Also squash it to one commit?

nvdv commented:
Done.

moutai committed:
Add Brier_score_loss to model validation doc

moutai committed:
Fix small typos in docstring and comment (#6840)

martinbjeldbak committed:
Typo documentation fixes in Advanced Installation (#6753)

This file could use a lot more clarity fixes...

Shorter line

Fix typo in "install_by_distribution"

bernhard2202 committed:
DOC - improved doc on ridges alpha parameter (#6844)

* improved doc on ridges alpha parameter

* updated alpha doc for all ridge classes

* fixed grammar

raghavrv commented:
@glouppe @ogrisel Could you review this one?

pianomania committed:
Add description about default score_func

kmike committed:
fixed error message for invalid class_weight values

jnothman committed:
Merge pull request #6850 from kmike/class_weight_error

fixed error message for invalid class_weight values

jnothman committed:
[MRG+1] FIX support memmap scalars as CV scores (#6789)

* FIX support memmap scalars as CV scores

* FIX test for Python 3.5 and NumPy 1.12

krishnakalyan3 committed:
fix balance

tguillemot committed:
Modification of the GaussianMixture class. (#6824)

Remove the useless computation of the covariance matrix.
Prepare the integration of the BayesianGaussianMixture class.

krishnakalyan3 committed:
fix Indentation and redundant info

krishnakalyan3 committed:
fix Indentation and redundant info

samuelstjean committed:
Doc: Fix typo (#6854)

jnothman committed:
Merge pull request #6853 from krishnakalyan3/class_weight_balance

DOC Update residual references of class_weight as balanced

viirya committed:
[MRG+1] Add a set recording pushed elements in DBSCAN (#6799)

* Add a push map recording pushed elements.

* Use set instead of map.

* Documented the p parameter.

ogrisel committed:
FIX workaround bug in numpy 1.9 (#6856)

Also use dtype with explicit precision (as a good practice) although
this does not impact the outcome of this test.

jnothman committed:
DOC insert missing function description

Otherwise sphinx latexpdf build breaks here

jnothman committed:
TST ensure tests are importable

jnothman committed:
DOC fix broken image path

jnothman committed:
PEP8

CaMeLCa5e committed:
Fixed pep8 errors

linting

CaMeLCa5e committed:
pep8

fixed whitespace

agramfort committed:
fix pep8

denadai2 committed:
Lasso docs fixed (#6880)

tguillemot committed:
Change GMM -> GaussianMixture. (#6886)

amueller commented on sklearn/tree/export.py:
```diff
@@ -415,3 +416,108 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        except ImportError:
+            raise ImportError(
+                'Plotting to Jupyter Notebook requires Jupyter Notebook.')
+        return Image(image, height=height, width=width)
```
you can actually just return ``src`` and it will show correctly.

nvdv replied:
That's true, but then there won't be any way to set image height/width

tguillemot replied:
In fact, they will have multiple solution to manage height/width with their favorite library.

amueller commented on sklearn/tree/export.py:
```diff
@@ -415,3 +416,108 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        filled=filled, leaves_parallel=leaves_parallel, impurity=impurity,
+        node_ids=node_ids, proportion=proportion, rotate=rotate,
+        rounded=rounded, special_characters=special_characters)
+    src = graphviz.Source(in_memory_dot_file.getvalue())
```
I might possibly make the part up to here it's own function, even if it's relatively short. That's all that's needed in the notebook. And always passing ``out="jupyter"`` seems a bit annoying. You could always return an object that knows how to render itself in jupyter. We haven't done any jupyter specific stuff in sklearn yet. I'm +1 on doing that but we might want to discuss that more broadly.

nvdv replied:
Acknowledged.

amueller commented on sklearn/tree/export.py:
```diff
@@ -415,3 +416,108 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        raise ImportError('Plotting requires graphviz to be installed.')
+
+    in_memory_dot_file = six.StringIO()
+    export_graphviz(
```
I think we should add a lower-level function that just creates the source and returns it. or add an option to ``export_graphviz`` to return just the source. Creating an IO object here is pretty unnecessary.

raghavrv replied:
Indeed which is why I suggest having `export(out={None|'<file_path>'}, format={'image'|'dot'})`. WDYT?

raghavrv replied:
Where `None` returns and `<file_path>` writes to file

and 'image' returns/writes to file the image object returned by `graphviz.Source(...).pipe(format='')` and 'dot' returns/writes to file the dot formatted graph string.

nvdv replied:
If I understand this correctly that means `export_graphviz` refactoring?

raghavrv replied:
Yes. Keep the function name to `export_graphviz` until someone approves of the name change. Currently I think you should add the `format={'image'|'dot'}` option, extend the functionality of `out={None|'<file_path>'}` to return the dot string or the image object (from src) if out is None and write it to a file if out is a valid filename.

tguillemot commented:
>I think we should add a lower-level function that just creates the source and returns it. or add an option to export_graphviz to return just the source. Creating an IO object here is pretty unnecessary.

For example, if you use matplotlib, subplot is not working as expected. So +1 for @amueller said.

krishnakalyan3 committed:
link to gpr guide (#6885)

xyguo committed:
add doctest extension (#6774)

amueller committed:
DOC add betaworks testimonial (#6825)

agramfort committed:
Update ISSUE_TEMPLATE.md

GaelVaroquaux committed:
Merge pull request #6613 from nelson-liu/correct_license_spelling

[MRG+1] DOC: fix spellings of 'license'

raghavrv committed:
[MRG] ENH Restructure grid_scores_ into a dict of 1D arrays

Also, DOC what's new section for model_selection changes

raghavrv committed:
CI Install psutil (#6897)

tguillemot committed:
[MRG] ENH Add Calinsky-Harabaz and Fowkes-Mallows clustering metrics (#6823)

Based on the code of A. Fouchet in PR#4301.

fishcorn committed:
Make KernelCenterer a _pairwise operation

Replicate solution to https://github.com/scikit-learn/scikit-learn/commit/9a520779c233dfeff466870c0b7cb04b705e61af except that `_pairwise` should always be `True` for `KernelCenterer` because it's supposed to receive a Gram matrix. This should make `KernelCenterer` usable in `Pipeline`s.

Happy to add tests, just tell me what should be covered.

lesteve committed:
Fix CircleCI failure (#6903)

The way the environment was activated was a la virtualenv. Maybe it used
to work for some time but not anymore.

[doc build]

nvdv committed:
Refactor export_graphviz function.

nvdv committed:
Refactor export_graphviz function.

nvdv commented:
@amueller @tguillemot @raghavrv 
I refactored `export_graphviz` function and created function that returns just the source. 
At present tree can be plotted by wrapping resulting source: `graphviz.Source(tree_source)`
Tell me what you think and I'll update comments and tests.

fishcorn committed:
Adding test for PR #6900

fishcorn committed:
Simplifying imports and test

jamoque committed:
updating changelog links on homepage (#6901)

rombiee committed:
first commit

rombiee committed:
changed binary average back to macro

rombiee committed:
changed binomialNB to multinomialNB

Commit:
emphasis on "higher return values are better..." (#6909)

b-carter committed:
fix typo in comment of hierarchical clustering (#6912)

yenchenlin committed:
[MRG] Allows KMeans/MiniBatchKMeans to use float32 internally by using cython fused types (#6846)

lesteve committed:
Fix sklearn.base.clone for all scipy.sparse formats (#6910)

jnothman committed:
DOC If git is not installed, need to catch OSError

Fixes #6860

jnothman committed:
DOC add what's new for clone fix

ryanyu9 committed:
fix a typo in ridge.py (#6917)

fishcorn committed:
pep8

GaelVaroquaux committed:
Merge pull request #6908 from HashCode55/master

[MRG+1] Changed the metric average types

GaelVaroquaux committed:
Merge pull request #6900 from fishcorn/fishcorn-patch-1

[MRG+1] Make KernelCenterer a _pairwise operation

GaelVaroquaux committed:
TST: Speed up: cv=2

This is a smoke test. Hence there is no point having cv=4

ClimbsRocks committed:
DOC Minor grammar fix (#6925)

crodjer committed:
DOC Fix text data tutorial (#6923)

 - Fix a typo.
 - Fix a floating point error in doctests.
 - Fix `VisibleDepricationWarning` due to conversion of an array with
   ndim > 0 to an index.
 - Convert `print` to `log` to avoid doctests failure.

Fixes: #6927

Signed-off-by: Rohan Jain <crodjer@gmail.com>

lesteve committed:
COSMIT fix typo in author name

imaculate committed:
[MRG+1] Added support for sample_weight in linearSVR, including tests and documentation. Fixes #6862 (#6907)

* Make KernelCenterer a _pairwise operation

Replicate solution to https://github.com/scikit-learn/scikit-learn/commit/9a520779c233dfeff466870c0b7cb04b705e61af except that `_pairwise` should always be `True` for `KernelCenterer` because it's supposed to receive a Gram matrix. This should make `KernelCenterer` usable in `Pipeline`s.

Happy to add tests, just tell me what should be covered.

* Adding test for PR #6900

* Simplifying imports and test

* updating changelog links on homepage (#6901)

* first commit

* changed binary average back to macro

* changed binomialNB to multinomialNB

* emphasis on "higher return values are better..." (#6909)

* fix typo in comment of hierarchical clustering (#6912)

* [MRG] Allows KMeans/MiniBatchKMeans to use float32 internally by using cython fused types (#6846)

* Fix sklearn.base.clone for all scipy.sparse formats (#6910)

* DOC If git is not installed, need to catch OSError

Fixes #6860

* DOC add what's new for clone fix

* fix a typo in ridge.py (#6917)

* pep8

* TST: Speed up: cv=2

This is a smoke test. Hence there is no point having cv=4

* Added support for sample_weight in linearSVR, including tests and documentation

* Changed assert to assert_allclose and assert_almost_equal, reduced the test tolerance

* Fixed pep8 violations and sampleweight format

* rebased with upstream

proinsias committed:
Fix minor typo in grid_search.py

glouppe committed:
Merge pull request #6935 from proinsias/proinsias-gs-typo

Fix minor typo in grid_search.py

jstriebel committed:
documentation: fixing typo (#6938)

the documentation of the form of regular and nested parameters in set_params is corrected

moutai committed:
fix pipeline docs typos (#6940)

markbao committed:
Doc: spectral.py: Add default value to gamma (#6941)

shotat committed:
Fix typos (#6942)

Commit:
DOC Add missing parameter defaults for some ensembles

glouppe committed:
Merge pull request #6944 from proinsias/proinsias-rfc-doc-defaults

DOC Add missing parameter defaults for some ensembles

proinsias committed:
Fix doc refs to StratifiedKFold and KFold (#6936)

Fix links to `StratifiedKFold` and `KFold` in docstrings.

proinsias committed:
DOC Add missing parameter defaults for Naive Bayes classes

TomDLT committed:
[MRG+2] convert to boolean arrays for boolean distances (#6932)

for example jaccard.

bnaul committed:
Properly handle bounds in isotonic regression with `np.clip` (#6949)

GaelVaroquaux committed:
Merge pull request #6946 from proinsias/proinsias-doc-nb

DOC Add missing parameter defaults for Naive Bayes classes

nvdv commented:
any news here? Thanks!

rasbt committed:
rename function to func (#6700)

jnothman committed:
TST avoid imports from deprecated module

Commit:
DOC make confusion_matrix_plot.py exmaple more copy-pastable

imaculate committed:
[MRG] Added sample weight parameter to LinearSVC.fit (#6939)

jnothman committed:
DOC what's new entries for LinearSVR/C weight support

ClimbsRocks committed:
Minor grammar fix

GaelVaroquaux committed:
Merge pull request #6979 from ClimbsRocks/patch-8

Minor grammar fix

Commit:
Fix minor typo

glouppe committed:
Merge pull request #6982 from fisache/master

Fix minor typo

vharavu committed:
update for issue #6523 (add logistic_regression_path to classes.rst (#6970)

* update for issue #6523 (add logistic_regression_path to classes.rst

* Clarifies statement about no warm-start and no speedup with liblinear

perimosocordiae committed:
avoid creating a diagonal matrix (#6989)

Broadcasted elementwise multiplication with a vector is
equivalent to matrix multiplication against a diagonal matrix,
but more efficient.

amueller committed:
Merge pull request #6966 from doshyt/doc-update-confusion

[MRG] DOC make confusion_matrix_plot.py example more copy-pastable

nicholas-fwang committed:
Fix minor typo in doc (#7001)

jannab committed:
[MRG + 1] docs: Change example of KDTree to prevent DeprecationWarning (#6962)

* docs: Change example of KDTree to prevent DeprecationWarning

- fill in missing import statement of sklearn.neighbors.KDTree
- wrap X in a list

Changes to be committed:
	modified:   sklearn/neighbors/binary_tree.pxi

Resolves: #6906

* docs: remove import statement of KDTree in example of KDTree

- remove import statement of sklearn.neighbors.KDTree

Changes to be committed:
	modified:   sklearn/neighbors/binary_tree.pxi

Resolves: #6906

gojomo committed:
[MRG+1] test against single-class y_true values for #6980 (#6987)

* test against single-class y_true values

* allow single label

btabibian committed:
removing confusing comment (#7007)

wyseguy7 committed:
Updated README, removed git clone with write reference (#7009)

jrfiedler committed:
Added deprecation cycle and modified examples (#7013)

* Added deprecation cycle and modified examples

* Removed newline and added clarifying example

willingc committed:
Edit CONTRIBUTING doc (#7017)

kjacks21 committed:
DOC Fixed check_estimator example (#7025)

copy & pasted example from scikit-learn-contrib template project

shengxiduan committed:
adding user guide link  to LinearDiscriminantAnalysis issue #6829 (#7031)

ogrisel committed:
Joblib 0.10.0 (#6985)

* joblib 0.10.0

* LC_ALL=C is necessary for sed

dhanus committed:
[MRG+1] Deprecated GMM functions (#7012)

* added @deprecated to gmm functions

* add lmvn deprecation tests

* updated exp_msg in test_lvmpdf_full_cv_non_positive_definit

* fixed deprecated warning messages in gmm

* fixed deprecated warning messages in tests

* fixed version numbers in deprecation statements.

* fix typo in warning message

* actually fixed the deprecation messages

* remove errant var and use spherecv

* rearrange tests

blakeflei committed:
[MRG + 1] update test_common.py deprecation warning (#7024)

* update test_common.py deprecation warning

* Updated test_common.test_get_params_invariance to omit deprecation warnings on deprecated classes per issue 7006

juliathebrave committed:
[MRG +1] adds DeprecationWarning to DPGMM and VBGMM (#7020)

* added Deprecation Warnings to DPGMM and VBGMM classes and other functions

* modified/added functions test_digamma, test_gammaln, test_log_normalize to check for DeprecationWarning

* added functions to test for Deprecated Warnings for wishart_log_det, wishart_logz

* updated test_log_normalize function

* fixed spelling in DeprecationWarning for class DPGMM and
added DeprecationWarning tests test_DPGMM_deprecation, test_VBGMM_deprecation

* added blank lines near inserted DeprecatedWarnings to conform to pep8

* added blank lines and deleted trailing whitespace for pep8

* removed 'import warnings' since not used

* changed wording of DeprecationWarnings for style and/or recommend alternate functions

* fixed spelling error in warning for gammaln

* suppress DecprecationWarnings for two functions

* fixed ignore_warnings function usage

* added more @ignore_warnings(DeprecationWarning)

* temporary commit to check nosetests on travis

* TST suppresses more DeprecatedWarnings

addresses GH6998 VBGMM and DPGMM deprecation

* CLN fixes whitespace for pep8

in test_gmm.py and test_dpgmm.py
for GH6998

* TST uncommented some code

I had commented some code for debugging purposes and had forgotten
to uncomment it. Now it is uncommented again.

nelson-liu committed:
DOC: fix spelling of Schütze in tfidf citations

jnothman committed:
DOC fix misconception in DBSCAN documentation

Fixes #6192

samuelstjean committed:
DOC: fix typo (#7044)

Small fix in the comments.

nelson-liu committed:
DOC: fix typo in sprint title in about (#7043)

Commit:
DOC update refs for MiniBatchKmeans

NickleDave committed:
typofix in modules/grid_search.rst (#7045)

Commit:
DOC fix MiniBatchKmeans See Also and colon

Commit:
DOC harmonize KMeans docstring with doc standards

rasbt committed:
remove redundant NotFittedError import from voting classifier (#7060)

neo01124 committed:
Typo

Changed 'will is to ' to 'will tend towards ', keeping in line with how the next line is written.

glouppe committed:
Merge pull request #7062 from neo01124/master

Typo

samuelstjean committed:
DOC: fix typo

Seems like I forgot some from the other day, all fixed now (according to the search bar at least).

NelleV committed:
Merge pull request #7063 from samuelstjean/patch-1

DOC: fix typo

Erotemic committed:
Fixed libsvm call with unicode kernel.
Added corresponding test.

Commit:
DOC fix typo in KMeans docstring

NelleV committed:
Merge pull request #7057 from doshyt/doc-update-confusion

[MRG] DOC update refs for MiniBatchKmeans

LeonieBorne committed:
[MRG+1] RadiusNeighborsClassifier edit (#6918)

* Bug with RadiusNeighborsClassifier 

Change 'weights' to 'weights[inliers]' on line 377

* Bug Radius Neighbors Classifier

Add test_radius_neighbors_classifier_with_precomputed_metric_and_outliers()

* Update test_neighbors.py

* Update test_neighbors.py

* Update whats_new.rst

* PEP8 issues

agramfort committed:
pep8

nelson-liu committed:
[MRG+3] Add mean absolute error splitting criterion to DecisionTreeRegressor (#6667)

* feature: add initial node_value method

* testing code for node_impurity and node_value

This code runs into 'Bus Error: 10' at node_value final assignment.

* fix: node_value now correctly calculating weighted median for sorted data.

Still need to change the code to work with unsorted data.

* fix: node_value now correctly calculates median regardless of initial order

* fix: correct bug in calculating median when taking midpoint is necessary

* feature: add initial version of children_impurity

* feature: refactor median calculation into one function

* fix: fix use of DOUBLE_t vs double

* feature: move helper functions to _utils.pyx, fix mismatched pointer type

* fix: fix some bugs in children_impurity method

* push a debug version to try to solve segfault

* push latest changes, segfault probably happening bc of something in _utils.pyx

* fix: fix segfault in median calculation and remove excessive logging

* chore: revert some misc spacing changes I accidentally made

* chore: one last spacing fix in _splitter.pyx

* feature: don't calculate weighted median if no weights are passed in

* remove extraneous logging statement

* fix: fix children impurity calculation

* fix: fix bug with children impurity not being initally set to 0

* fix: hacky fix for a float accuracy error

* fix: incorrect type cast in median array generation for node_impurity

* slightly tweak node_impurity function

* fix: be more explicit with casts

* feature: revert cosmetic changes and free temporary arrays

* fix: only free weight array in median calcuation if it was created

* style: remove extraneous newline / trigger CI build

* style: remove extraneous 0 from range

* feature: save sorts within a node to speed it up

* fix: move parts of dealloc to regression criterion

* chore: add comment to splitter to try to force recythonizing

* chore: add comment to _tree.pyx to try to force recythonizing

* chore: add empty comment to gradient boosting to force recythonizing

* fix: fix bug in weighted median

* try moving sorted values to a class variable

* feature: refactor criterion to sort once initially, then draw all samples from this sorted data

* style: remove extraneous parens from if condition

* implement median-heap method for calculating impurity

* style: remove extra line

* style: fix inadvertent cosmetic changes; i'll address some of these in a separate PR

* feature: change minmaxheap to internally use sorted arrays

* refactored MAE and push to share work

* fix errors wrt median insertion case

* spurious comment to force recythonization

* general code cleanup

* fix typo in _tree.pyx

* removed some extraneous comments

* [ci skip] remove earlier microchanges

* [ci skip] remove change to priorityheap

* [ci skip] fix indentation

* [ci skip] fix class-specific issues with heaps

* [ci skip] restore a newline

* [ci skip] remove microchange to refactor later

* reword a comment

* remove heapify methods from queue class

* doc: update docstrings for dt, rf, and et regressors

* doc: revert incorrect spacing to shorten diff

* convert get_median to return value directly

* [ci skip] remove accidental whitespace

* remove extraneous unpacking of values

* style: misc changes to identifiers

* add docstrings and more informative variable identifiers

* [ci skip] add trivial comments to recythonize

* remove trivial comments for recythonizing

* force recythonization for real this time

* remove trivial comments for recythonization

* rfc: harmonize arg. names and remove unnecessary checks

* convert allocations to safe_realloc

* fix bug in weighted case and add tests for MAE

* change all medians to DOUBLE_t

* add loginc allocate mediancalculators once, and reset otherwise

* misc style fixes

* modify cinit of regressioncriterion to take n_samples

* add MAE formula and force rebuild bc. travis was down

* add criterion parameter to gradient boosting and add forest tests

* add entries to what's new

sshleifer committed:
DummyClassifier and DummyRegressor should raise NotFittedError (#7069)

arcolife committed:
fixes #6829 - user guide links to DiscriminantAnalysis (#6838)

lesteve committed:
Merge pull request #7041 from jnothman/dbscan_doc

[MRG + 1] DOC fix misconception in DBSCAN documentation

jrfiedler committed:
DOC Reword description to remove ambiguity (#7066)

lesteve committed:
DOC joblib default start method is not forkserver (#7077)

for Python 3.4+. This was reverted in joblib 0.9.3 in order to support
interactive functions passed into joblib.Parallel.

NelleV committed:
Merge pull request #7033 from nelson-liu/fix_tfidf_intro_to_ir_citation

DOC: fix spelling of Schütze in tfidf citations

yenchenlin committed:
COSMIT Reduce duplicated code (#7053)

raghavrv committed:
MNT use model_selection (#7082)

amueller committed:
DOCS / COSMIT duplicate word typos

NelleV committed:
Merge pull request #7083 from amueller/typos

DOCS / COSMIT duplicate word typos

jnothman committed:
DOC revert removal of duplicate word in doctest

amueller committed:
add __dir__ to bunch for better autocomplete (#7090)

agramfort committed:
pep8

nelson-liu committed:
[MRG+1] feature: add beta-threshold early stopping for decision tree growth (#6954)

* feature: add beta-threshold early stopping for decision tree growth

* check if value of beta is greater than or equal to 0

* test if default value of beta is 0 and edit input validation error message

* feature: separately validate beta for reg. and clf., and add tests for it

* feature: add beta to forest-based ensemble methods

* feature: add separate condition to determine that beta is float

* feature: add beta to gradient boosting estimators

* rename parameter to min_impurity_split, edit input validation and associated tests

* chore: fix spacing in forest and force recompilation of grad boosting extension

* remove trivial comment in grad boost and add whats new

* edit wording in test comment / rebuild

* rename constant with the same name as our parameter

* edit line length for what's new

* remove constant and set min_impurity_split to 1e-7 by default

* fix docstrings for new default

* fix defaults in gradientboosting and forest classes

nateGeorge committed:
Fix docs links (#7005)

* docs: fix broken and redirect links

see #7000

* docs: fix links

see #7000

* docs: merge with master

* docs: fix fnrs and tinyclues logo links

* docs: fix link reference in text

* docs: fix typo

* docs: added back in metaoptimize-qa paragraph

* docs: update language for defunct site

* docs: update stackexchange section

* docs: remove defunct site, move quora to top

* docs: remove defunct link and rearrange links

amueller committed:
add missing verb in whatsnew

Titan-C committed:
Sphinx-doc absolute path for auto_example images

staubda committed:
Updated docstring for bootstrap argument of IsolationForest (#7095)

jrfiedler committed:
Modify v0.18 DeprecationWarning messages (#7067)

nelson-liu committed:
add versionadded for MAE (#7105)

nelson-liu committed:
DOC: fix spacing in forest mae versionadded (#7106)

nelson-liu committed:
[MRG + 1] add "versionadded" tags to min_impurity_split docstrings (#7093)

* add versionadded tags to min_impurity_split docstrings

* adjust positioning of versionadded tags

* remove spurious comment used to force recythonization

ixjlyons committed:
[MRG+1] Check for bad labels in confusion_matrix. (#6958)

* Check for bad labels in confusion_matrix.

* Raise exception if all specified labels are not in y_true.

* Use np.intersect1d instead of membership checking in list comprehension.

* Revert back to using list comprehension.

nelson-liu committed:
doc: fix early stopping indentation in what's new (#7108)

rammig committed:
[MRG+1] Added text and logo for ottogroup testimonial (#6802)

* Added text and logo for ottogroup testimonial

* fixed typo

* fixed typo

Ajackster committed:
[MRG+1] DOC Modified the navbar menu to be responsive with mobile. (#7098)

* Modified the navbar menu to be responsive with mobile.

* Gave the navbar a fixed height

* Gave more padding to the hamburger

* Fixed text overflow issue on mobile menu

* Got rid of extra line

* Made the whole hamburger clickable

manu-chroma committed:
[MRG + 1] load_iris dataset: added return_X_y option (#7049)

* load_iris dataset:added return_X_y option

* Updated return type description

* improved return type description

* Removed extra line

* Added extra line

* Added test sentence.

* Remove sample text

* Fixes

* pep8

* Added version tag

* added entry in whats_new

* fixed extra space

jrfiedler committed:
Removed unused imports

jnothman committed:
[MRG + 1] ENH show no warning with chi2 of empty feature (#6978)

* ENH show no warning with chi2 of empty feature

* TST better error message when warnings raised

* TST fix test in old Numpy where another warning is issued

GaelVaroquaux committed:
Merge pull request #7064 from Erotemic/quickfix_py2_libsvm_kernel_unicode

Fixed python2 libsvm call with unicode kernel

GaelVaroquaux committed:
Merge pull request #7119 from jrfiedler/remove-unused-imports

[MRG] Removed unused imports

lesteve committed:
COSMIT fix python2 -> python3 in comment

Commit:
ENH TfidfTransformer.transform() performance bug fix (#7125)

* Fix performance bug in TfidfTransformer.transform()

save self._idf_diag as csr_matrix to avoid transforming self._idf_diag to csr_matrix each transform() call.
It happens when self.use_idf==True in: X = X * self._idf_diag in __mul__ operator.

olologin committed:
[MRG] fix #5269:  Overflow error with sklearn.datasets.load_svmlight (#7101)

* fix for #5269, overflow error

* test with long qid added

* What's new section added

agramfort committed:
DOC rm obsolete DATASET_PROPOSAL.txt (#7133)

raghavrv committed:
[MRG] MAINT Removing more deprecated stuff for v0.18 (#5528)

* MNT Removing more deprecated stuff for v0.18

* MNT+TST Remove support for loss='l2'

* MNT Remove support for precompute='auto' for ElasticNet

* FIX/TST precompute=auto should raise a generic message

GaelVaroquaux committed:
Merge pull request #4658 from Titan-C/sphx-glr

[MRG+2] Sphinx-gallery use

GaelVaroquaux committed:
Contributing statement (#7097)

* MISC: remove broken link

* Statement on the process + values of the community

To answer the demand for a CoC

* DOC: add links to contributing page

* DOC: Documenation contribution are important

JulianKahnert committed:
fix of typo (#7137)

nelson-liu committed:
fix indent in whats_new and fix indent in datasets/base.py (#7138)

iver56 committed:
Fix typo in example feature_stacker.py

yenchenlin committed:
Fix typo (#7145)

naoyak committed:
add load_breast_cancer to API reference (#7149)

tguillemot committed:
Correct Deprecation of DPGMM and VBGMM. (#7124)

nelson-liu committed:
fix MultiOutputEstimator fit return docstring (#7150)

nelson-liu committed:
fix some sphinx warnings in whats_new.rst (#7151)

manu-chroma committed:
[MRG] load_breast_cancer dataset: added return_X_y option (#7152)

* added return_X_y option to breast_cancer_dataset, tests included

* fix typo

* update whats_new

* removed extra space

* improved whats_new changelog

staubda committed:
[MR] Speed issues in sample_without_replacement - bugfix (#7104)

* Remove cdivision=True Cython compiler directive causing incorrect int division.

* Decrease tol for LogisticRegression fit test, perhaps needed due to test not controlling `random_state`.

manu-chroma committed:
[MRG+1] added return_X_y option to toy datasets in sklearn.datasets (#7154)

* added return_X_y support to more dataset loaders

* fix typo

* updated whats_new.rst

* fix indentation for version added tag

* call astype before the branching

* better formatting in whats_new.rst

* better formatting

* updated what's new

tguillemot committed:
[MRG+1] Dpgmm alpha (#7028)

* BF: not mutating alpha in VBGMM, using alpha_ instead

* Correct alpha_.

agramfort committed:
cosmit

btabibian committed:
fixing ElasticNet sparse decision function to match output with dense output

btabibian committed:
test case for ElasticNet multiple output matching with dense

btabibian committed:
pushing Transpose into dot product

btabibian committed:
removing make_sparse_data call

btabibian committed:
removing extra fit call and cleanup code

Commit:
fixed typo

GaelVaroquaux committed:
Merge pull request #7165 from tKammy/fixed_typo

fixed typo

btabibian committed:
droping extra y

btabibian committed:
fixing alignment

yenchenlin committed:
DOC Add LabelKFold in _BaseKFold's docstring (#6344)

lesteve committed:
Use pgtol parameter in fmin_l_bfgs call (#7167)

in HuberRegressor. Since fmin_l_bfgs does not take a 'tol' parameter it
resulted in the scipy < 0.9 branch always being called.

tguillemot committed:
[MRG+2] Modification of GaussianMixture class. (#7123)

* Modification of GaussianMixture class.

The purpose here is to prepare the integration of BayesianGaussianMixture.

* Fix comments.

* Modification of the Docstring.

* Add license and author.

* Fix review and add tests for init.

ogrisel committed:
TST make test_forest robust to platform variability

Under linux 32 bit test_boston fails because the score is slightly
below 0.95.

This could come from the fact the upcasting from float to double variables in C
(generated from Cython code in this case) can yield different values when
compiled to run under different bitness of the Linux runtime environment.
Those small discrepancies can accumulate when computing the criterion or
threshold values.

This part of the Cython code is not enough unittested to make it easy to
exactly pinpoint the exact location where this happens though.

jessicayung committed:
DOC Typo: consequective -> consecutive (#7171)

betatim committed:
[MRG] Check for NaN or Inf in metric input (#6976)

jasonlaska committed:
Add spherecluster related package

jasonlaska committed:
Address reviewer comments

I changes the text to "data on the unit hypersphere."  Hopefully that sounds less awkward (but is also precise).

yenchenlin committed:
[MRG] Make Silhouette_score support sparse X (#7170)

ogrisel committed:
FIX broken NMF test with numpy 1.7.1

ogrisel committed:
STYLE trailing whitespace

ogrisel committed:
FIX broken iforest test under numpy 1.7.1

Also improve error message to be more informative.

GaelVaroquaux committed:
Merge pull request #7172 from jasonlaska/patch-1

[MRG+1] Add spherecluster related package

Commit:
[MRG] extending tfidf documentation (#7015)

superbobry committed:
BUG fixed a typo in ``sklearn.gmm`` (#7182)

``_log_multivariate_normal_density_spherical`` failed with ``IndexError``
when ``covar`` was a scalar because of the typo in shape check.

Commit:
[MRG+2] ENH: (minor) add quantile_low/high parameters to robust scaler (#5929)

* add quantile_low/high parameters to robust scaler

add versionadded tags

* robust scaler: add parameter for quantile_range, check its validity

* robust scaler: simplify params validations

fix typo in docstrings of RobustScaler

* preprocessing.data fix some lines > 80 columns

* add whatsnew for robust scaler quantile_range param

add missing url link in whatsnew

tguillemot committed:
Fix -np.infty in GaussianMixture. (#7180)

agramfort committed:
cosmit

srvanrell committed:
Doc minor fix ensemble.rst (#7186)

Remove repeated "at each"

yenchenlin committed:
[MRG] Add whats_new about fixing example (#7184)

* Add whats_new

* Fix typo

* Fix wording

agramfort committed:
rephrasing

jnothman committed:
DOC fix Olivier's misspelt name

nelson-liu committed:
[MRG] edit citation indentation in about.rst and add gsoc info for '15 '16 (#7191)

GaelVaroquaux committed:
Merge pull request #7143 from iver56/patch-1

Fix typo 'to' => 'too'

ogrisel committed:
FIX broken test when run with arch -i386 on OSX

yenchenlin committed:
[MRG+1] Rename CV params n_{folds,iter} to n_splits (#7187)

* Rename n_iter to n_splits

* Fix bug

* Fix examples

* Add spaces

* Rename n_folds to n_splits

* Fix error

* Fix doc

* Fix doc

* Fix example

* Rename variables name

* PEP8

* Fix error message

* Add whats_new

* Fix test

* Fix doc

* Fix doc

* Make test clear

ceshine committed:
DOC Correction: max_depth is effective when max_leaf_nodes is not None (#7183)

jnothman committed:
DOC improve formatting for learning_rate docstrings

MechCoder committed:
[MRG + 1] [MAINT] Remove non-trivial logic during kernel initialization (#6963)

* Make sure set_params does not break clone on kernels

For example, in kernels such as RBF, using set_params with a
length_scale of different length, used to break clone because of
the non-trivial modification of length_scale in the constructor.

This PR resolves this by moving this logic to __call__ and
making hyperparameter attributes properties.

* Fix tests by using get_params and implement __eq__ for Hyperparameter

* Explain XXX

* Check that hyperparameters are equal after non-trivial set_params

* Refactor into _check_length-scale

ogrisel committed:
FIX make HuberRegressor tests stable under 32 bit Linux (#7196)

definitelyuncertain committed:
Explicit warnings in model persistence documentation (#7157)

* Update model_persistence.rst

Mention a reason not to unpickle untrusted data.

Be more precise about using pickled data from different versions.

* Explicit warning about cross version pickling

kjacks21 committed:
[MRG] DOC updated normalize param docstring (#7161)

* DOC updated normalize param docstring

* DOC corrected StandardScaler reference

* DOC changes to alpha param docstring

* DOC :class:

nelson-liu committed:
clarify splitting sparse data always outputs csr matrix

nelson-liu committed:
edit docstring for arrays argument as well

nelson-liu committed:
remove accidentally added whitespace

nelson-liu committed:
removed versionadded for arrays param and fixed spacing

caioaao committed:
[MRG+1] Fix for "_transform_selected" not copying when "selected='all'" (#7201)

* fixes issue scikit-learn/scikit-learn#7194

* Added test

* Making `selected='all'` explicit on test

* Updated whats_new.rst

* Fixed typo on `whats_new.rst`

nelson-liu committed:
feature: add tests for sparse matrices

nelson-liu committed:
try to fix travis error

nelson-liu committed:
be more explicit about purpose of test

GaelVaroquaux committed:
Merge pull request #7199 from nelson-liu/edit_cv_sparse_docs

[MRG+1] DOC: Clarify splitting sparse data always outputs csr matrix

zhengruifeng committed:
GaussianNB now support sample weighting

GaelVaroquaux committed:
Merge pull request #7213 from zhengruifeng/doc_fix

Fix #7212  Sync Document for GaussianNB

ClimbsRocks committed:
minor wording fix

Changes "the scikit" to "scikit-learn". While both might be technically accurate, explicitly calling the package "scikit-learn" helps with naming consistency.

GaelVaroquaux committed:
Merge pull request #7217 from ClimbsRocks/patch-9

minor wording fix

agramfort committed:
update what's new + simplify test

b-carter committed:
fix issues with manifold tests throwing pep8 errors (#7209)

nvdv commented:
any news here?

rasbt committed:
update setup.py PyPI trove classifier section (#7197)

* update pypi trove classifier section

* add 3.4 back

amueller commented:
why not just allow ``export_graphviz`` to return the string?

amueller commented:
I think the current approach is the right one.
But we also need tests (there is a test failure and conflicts right now) and the example doesn't conform to what the code does.

amueller commented:
I'm not sure about the current features of sphinx-gallery. Can we use the notebook display stuff in examples? (that would be kinda sweet ;)

GaelVaroquaux commented:
> I'm not sure about the current features of sphinx-gallery. Can we use the notebook display stuff in examples?

Are you talking about the notebook integration of display hooks (maybe
they have a better name)?

Sphinx-gallery does not do any of this: it's goal is not to reimplement
Jupyter :).

staubda committed:
[MRG] Remove heavy memory footprint in BaseBagging due to OOB scoring (#7118)

* Remove heavy memory footprint for OOB scoring from bagging.

- Remove `estimators_samples` attribute from `BaseBagging`
- Add method `_get_estimators_samples` to `BaseBagging` that
  returns a generator producing sample indices on demand.
- Slight refactor of `_parallel_build_estimators()` to
  isolate and group lines accessing random state.

* Replaced `BaseBagging.estimators_samples_` attribute with property.

- Accessing `BaseBagging.estimators_samples_` now triggers call to
  `BaseBagging._get_estimators_samples` as well as deprecation
  warning for release 0.20.
- This should make the fix for the OOB memory issue fully backwards
  compatible.

* Remove print statement.

* Added test, made new method more general.
- Added test to `test_bagging.py` to ensure that indices generated
  on the fly are identical to indices generated at fit time.
- Refactored `_get_estimators_samples()` to `_get_estimators_indices()`,
  now both feature and samples indices are returned.
- Refactored `estimators_samples_()` to deal with above.

* Rename index generating functions and arguments.

* `estimators_samples_()` returns list instead of generator

* Removed `estimators_samples_` deprection warning.

* Actually removed `estimators_samples_` deprecation warning.

* New mask generation function, new bagging test, new BaseBagging attributes.
- Added new private function in bagging.py that converts indices to
  a boolean mask.
- Added new bagging test to make sure identical OOB scores are generated
  when the same estimator if fit with fixed random state and identical
  training data.
- Added new private attributes `BaseBagging._max_features` and
  `BaseBagging._max_samples` to store validated input values.

* Streamlined code, improved `estimators_samples_()` documentation.
- Removed `max_samples` argument from `_parallel_build_estimators()`,
  this value is now accessed via `ensemble._max_samples`.
- Removed validation of `ensemble.max_features` and `max_samples`,
  instead use `ensemble._max_features` and `ensemble._max_samples`
  which are assumed to be already validated.
- Removed unnecessary `samples` variable from `_parallel_build_estimators()`.
- Changed the way `random_state` is generated in `_parallel_build_estimators()`
  and `BaseBagging._get_estimators_data_draws()` to direct numpy method to
  reflect that seeds created in `BaseBagging._fit()` are trustworthy.
- Due to above removed generation of new seed for each estimator in
  `_parallel_build_estimators()` and `BaseBagging._get_estimators_data_draws()`.
- Added documentation to `BaseBagging.estimators_samples_()` property
  indicating the reason it's generated dynamically and the associated performance
  penalty.
- Returned `BaggingClassifier._set_oob_score()` and `BaggingRegressor._set_oob_score()`
  to directly accessing `self.estimators_samples_`.

* Streamlined code, improved `estimators_samples_()` documentation.
- Removed `max_samples` argument from `_parallel_build_estimators()`,
  this value is now accessed via `ensemble._max_samples`.
- Removed validation of `ensemble.max_features` and `max_samples`,
  instead use `ensemble._max_features` and `ensemble._max_samples`
  which are assumed to be already validated.
- Removed unnecessary `samples` variable from `_parallel_build_estimators()`.
- Changed the way `random_state` is generated in `_parallel_build_estimators()`
  and `BaseBagging._get_estimators_data_draws()` to direct numpy method to
  reflect that seeds created in `BaseBagging._fit()` are trustworthy.
- Due to above removed generation of new seed for each estimator in
  `_parallel_build_estimators()` and `BaseBagging._get_estimators_data_draws()`.
- Added documentation to `BaseBagging.estimators_samples_()` property
  indicating the reason it's generated dynamically and the associated performance
  penalty.
- Returned `BaggingClassifier._set_oob_score()` and `BaggingRegressor._set_oob_score()`
  to directly accessing `self.estimators_samples_`.

* PEP8, remove optional arguments, add public function to utils.
- Update indentation and line length to conform to PEP8.
- Update `estimators_samples_()` docstring to conform to PEP8 and
  PEP257.
- Remove optional arguments to `BaseBagging._get_estimators_data_draws()`,
  rename to `BaseBagging._get_estimators_indices()`.
- Remove optional arguments from `_generate_mask_from_indices()`,
  move to `utils.metaestimators`, rename to `indices_to_mask()`.

* Add indices_to_mask to __all__ in utils.metaestimators

* Move utils function, new test for `estimators_samples` correctness.
- Moved `indices_to_mask` from `utils.metaestimators` to `utils.validation`
- Added new test to `ensemble.test_bagging` to make sure refitting
  an individual estimator from the bagging ensemble using the corresponding
  samples identified in `BaseBagging.estimators_samples` returns the same model.

* Adjust/remove tests, move `indices_to_mask` to `utils.__init__`.
- Remove test_bagging.test_consistent_index_sampling
- Update formatting checks in test_bagging.test_estimators_samples
  to make sure each mask is numpy boolean array
- Move indices_to_mask from utils.validation to utils.__init__

* update class docstring for `estimators_samples_`

- Updated class docstring of `estimators_samples_` attribute for `BaggingClassifier` and `BaggingRegressor` to indicate that samples are identified with boolean masks.

* PEP8, made `max_samples` arg optional in _fit()
- Fixed PEP8 issues.
- Gave `max_samples` argument of `BaseBagging._fit()` a default
  value of `None` as indicated in the docstring. Added code to
  check for and process `None` value.
- Added a few more comments to `_fit()`.

* 2 new `max_samples` consistency tests, update whats_new.rst

* Minor change

* Update whats_new.rst

nelson-liu committed:
Minor doc fixes in pairwise metrics (#7227)

nvdv commented:
> I think the current approach is the right one.
> But we also need tests (there is a test failure and conflicts right now) and the example doesn't conform to what the code does.
I just wanted to reach an agreement first and then fix tests and examples.

nvdv committed:
Fix tests, docs and examples for refatctored version of export_graphviz.

nvdv committed:
Merge branch 'tree_plotting' of github.com:nvdv/scikit-learn into tree_plotting

nvdv commented:
Oops. I accidentally did a rebase. Trying to clean that up...
